pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,215 @@
+package secret
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type SecretEventHandlerRegistration interface {
+	cache.ResourceEventHandlerRegistration
+
+	GetKey() ObjectKey
+	GetHandler() cache.ResourceEventHandlerRegistration
+}
+
+type SecretMonitor interface {
+	//
+	AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error)
+	//
+	RemoveSecretEventHandler(SecretEventHandlerRegistration) error
+	//
+	GetSecret(SecretEventHandlerRegistration) (*v1.Secret, error)
+}
+
+type secretEventHandlerRegistration struct {
+	cache.ResourceEventHandlerRegistration
+	// objectKey will populate during AddEventHandler, and will be used during RemoveEventHandler
+	objectKey ObjectKey
+}
+
+func (r *secretEventHandlerRegistration) GetKey() ObjectKey {
+	return r.objectKey
+}
+
+func (r *secretEventHandlerRegistration) GetHandler() cache.ResourceEventHandlerRegistration {
+	return r.ResourceEventHandlerRegistration
+}
+
+type secretMonitor struct {
+	kubeClient kubernetes.Interface
+
+	lock sync.RWMutex
+	// monitors is map of singleItemMonitor. Each singleItemMonitor monitors/watches
+	// a secret through individual informer.
+	monitors map[ObjectKey]*singleItemMonitor
+}
+
+func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
+	return &secretMonitor{
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*singleItemMonitor{},
+	}
+}
+
+// create/update secret watch.
+// routeSecretName is a combination of "routename_secretname"
+func (s *secretMonitor) AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
```

Comment:
Havne't made it through, but by is the route name necessary at all.  the secret namespace and name shoudl be sufficient.

Reply:
Thanks for the suggestion; I've done some refinement, and with the latest changes, the `secret_monitor` should work independent of route, and hence only need secret name and namespace